### ISSUE TITLE
[Feature request] It would be nice to be able to send some predefined data to program STDIN while fuzzing in persistent mode

### ISSUE THREAD

**Issue by nataraj-hates-MS-for-stealing-github** [AFLplusplus/AFLplusplus#567]
This feature request is follow up to [this thread](https://groups.google.com/forum/#!topic/afl-users/TX4DNXNpaH0), as as far as I can get, the feature I need is not available.

I am trying to fuzz some program in persitent mode. The program code is not changed, I added __AFL_LOOP inside extention.

To reach this __AFL_LOOP I need to send some predefined data to program's STDIN. From inside I will read fuzzed data from file whose specified in aflpp's -f option. But I need to send some data to STDIN to make program run this loop.

As far as I can get, it is not possible now.

This is short story. Now same story, but longer version.

I am trying to fuzz some functions from postgres internals. They are well available from extensions, so I wrote an extension with __AFL_LOOP inside.

Portgres has so called "single-user mode". In this mode postgres started with limited capabilities as single process. It is good to run fuzzing in single-user mode.

To run my __AFL_LOOP I had to issue an SQL command that will run function from my extension, with __AFL_LOOP in it.

In single-user mode, the only way to execute a command, is to send it to the STDIN of the postgres process.  Something like that, if we run from bash:

```
echo "select * from my_test('/home/nataraj/do/postgres/src/test/fuzzing/type_fuzzing/sample.txt');" | ASAN_OPTIONS=detect_leaks=0 /usr/local/pgsql/bin/postgres --single -D /tmp/aaa/ postgres 
```

So I need a way to send some predefined data while running afl++. Like

```
/usr/local/bin/afl-fuzz -i in -o out -t 10000 -f /tmp/my_data -m none -- "echo \"select * from my_test('/tmp/my_data');\" | ASAN_OPTIONS=detect_leaks=0 /usr/local/pgsql/bin/postgres --single -D /tmp/aaa/ postgres"
```

It would make my life much more easy...

**Comment by vanhauser-thc** [comment]
all targets I know only need one, file or stdin.
both I have never seen before, so unless this is something that comes up more often or cant be solved otherwise I would prefer not to add to the existing afl complexity.
If you already use a persistent mode harness, why dont you write write to the stdin file descriptor there?

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
> all targets I know only need one, file or stdin.

My target needs file.

But I also need to put some const-string text value  to the STDIN to make it work. This value should not be fuzzed, it is needed only to do proper "goto" to the part of the code where my __AFL_LOOP is.

> If you already use a persistent mode harness

I am not using harness. I am running the "full system" (or almost full) and then make it to "goto" to the __AFL_LOOP I need.And I need to send a "command" via STDIN to make this "goto"

**Comment by domenukk** [comment]
You should be able to solve this with an LD_PRELOAD, simply return your value on stdin read, similar to preeny

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
Can you please inspire me with a simple example, please. Or with link to similar solution. I can't grasp the idea from this suggestion. :-(

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
If you give me a good hint, I will be able to write a good guide for others, that meets same problem

**Comment by domenukk** [comment]
Read up on LD_PRELOAD, for example
http://www.goldsborough.me/c/low-level/kernel/2016/08/29/16-48-53-the_-ld_preload-_trick/

Ah and to use it in the afl fuzz target, use AFL_PRELOAD instead, so you don't redirect the fuzzer

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
I've written a lib setstdin in preeny stile, and offered preeny upstream to include it in the project.
https://github.com/zardus/preeny/issues/75
If they accept it, I will write a howto about using it from preeny.
If they reject it, I will write a howto using it as a standalone lib

**Comment by vanhauser-thc** [comment]
you can also send a PR here that adds it to our examples/ directory, e.g. "examples/send_stdin_fuzzing", look at argv_fuzzing and socket_fuzzing as examples.

**Comment by fuzzah** [comment]
@nataraj-hates-MS-for-stealing-github sorry for probably too late answer.
Please take a look at this article:
https://securitylab.github.com/research/fuzzing-challenges-solutions-1
 
They're fuzzing both ARGV and FILE. You could use same technique to achieve fuzzing of both STDIN and FILE.
TLDR: just assume that first N bytes in your input file are STDIN input and the rest part is actual input FILE. Surely it requires more modifications to code.
Hope this helps :)
 
upd: of course as your stdin string should always be the same you could just hardcode this string inside postgres main function or just make this 'goto' to your required location.
 
upd2: you could also try using afl-fuzz ENV variable AFL_SKIP_BIN_CHECK and starting afl-fuzz like this:
`env AFL_SKIP_BIN_CHECK=1 afl-fuzz -i in -o out -m none -- /bin/bash "echo 'MY INPUT STRING' | my_target @@"`
Not sure if afl-fuzz will replace those @ symbols with file name but I think it's manageable somehow with correct quotes combination.

**Comment by nataraj-hates-MS-for-stealing-github** [comment]
I've written a step by step guide how to pass STDIN to target binary via preeny setstdin

Now it is here https://gitlab.com/dhyannataraj/fuzzing-examples/-/wikis/How-to-pass-STDIN-data-to-target-binary-via-preeny-setstdin

Where should I PR it to make it easily available for anyone may need it?

Or should I put it here in local wiki? Where exactly?

**Comment by vanhauser-thc** [comment]
@nataraj-hates-MS-for-stealing-github  sorry for being very late on this.
IMHO this would fit best in docs/FAQ.md

**Comment by domenukk** [comment]
Closing as this is solved with LDPRELOAD and there is no activity.
